numerics/elliptic_functions.cpp

```diff
@@ -138,65 +137,63 @@ void JacobiSNCNDN(double const u,
                   double& s,
                   double& c,
                   double& d) {
-  double m, kc, ux, k, kh, kh3, kh5, kh7, k2, k3, k4, sx;
+  constexpr double k_over_2_lower_bound = π / 4.0;
 
-  m = 1.0 - mc;
-  kc = sqrt(mc);
-  ux = abs(u);
-  if (ux < 0.785) {
-    JacobiSNCNDNReduced(ux, mc, s, c, d);
+  // The argument reduction follows Fukushima (2009), Fast computation of
+  // Jacobian elliptic function and incomplete elliptic integrals for constant
+  // values of elliptic parameter and elliptic characteristic, sections 2.4 and
+  // 3.5.2.
+  double const m = 1.0 - mc;
+  double const kʹ = Sqrt(mc);
+  double abs_u = Abs(u);
+  if (abs_u < k_over_2_lower_bound) {
+    JacobiSNCNDNReduced(abs_u, mc, s, c, d);
   } else {
-    k = EllipticK(mc);
-    kh = k * 0.5;
-    kh3 = k * 1.5;
-    kh5 = k * 2.5;
-    kh7 = k * 3.5;
-    k2 = k * 2.0;
-    k3 = k * 3.0;
-    k4 = k * 4.0;
-    ux = ux - k4 * static_cast<double>(static_cast<int>(ux / k4));
-    if (ux < kh) {
-      JacobiSNCNDNReduced(ux, mc, s, c, d);
-    } else if (ux < k) {
-      ux = k - ux;
-      JacobiSNCNDNReduced(ux, mc, s, c, d);
-      sx = c / d;
-      c = kc * s / d;
+    double const k = EllipticK(mc);
+    double const k_over_2 = 0.5 * k;
+    double const three_k_over_2 = 1.5 * k;
+    double const two_k = 2.0 * k;
+    double const five_k_over_2 = 2.5 * k;
+    double const three_k = 3.0 * k;
+    double const seven_k_over_2 = 3.5 * k;
+    double const four_k = 4.0 * k;
```

Comment:
k/2, 3k/2, 5k/2, and 7k/2 are all used only once; inline them.

Consider trusting common subexpression elimination and inlining `2 * k`, `3 * k`, and `4 * k` as well.

Reply:
First part done.